### DISCUSSION TITLE
힙

### DISCUSSION THREAD

**Discussion by onee-only** [subak-ba/cs-study#22] [자료 구조]
# 힙 (Heap)
힙은 힙 프로퍼티(heap property)와 를 만족하는 트리 기반 자료 구조입니다.

> 힙 프로퍼티의 정의: 
어떠한 노드 C가 있을 때, 노드 C의 부모 노드 P의 값은 C의 값보다 작거나(최소 힙) 커야 한다(최대 힙). 

위 규칙에 따라 힙의 루트 노드는 최소 힙일 때 가장 작은 값, 최대 힙일 때 가장 큰 값이 됩니다. 
덕분에 힙의 (최소 or 최대)값을 O(1) 만에 찾을 수 있습니다.

아래는 최소 힙의 예시입니다.

![image](https://github.com/subak-ba/cs-study/assets/103808587/a128c289-252a-46fe-b07b-21e690217eee)

힙의 원소들은 사촌들(같은 깊이지만 부모가 다른 원소)과 일정한 대소관계가 성립하지 않습니다. 
위 자료에서 보실 수 있듯이, 힙 프로퍼티만을 만족하고 있습니다.

힙은 추상 자료형(abstract data type)인 _우선 순위 큐_ 의 구현체입니다. 
힙의 최대 혹은 최솟값이 언제나 루트 노드에 저장이 되어 있기 때문에, 어느 집합에서 지속적으로 최대 혹은 최솟값을 꺼내야 할 때 사용할 수 있습니다.

## 이진 힙 (Binary Heap)
가장 흔히 찾아볼 수 있는 힙의 구현체는 완전 이진 트리를 이용한 구현체입니다.

> 완전이진트리: 트리의 위부터 아래, 왼쪽부터 오른쪽의 순서로 빠짐없이 가득 차있는 이진 트리 (위 자료 사진 참고)

완전이진트리의 이러한 성질 덕분에 이진 힙은 배열로 표현하기 매우 좋은 구조입니다.

![image](https://github.com/subak-ba/cs-study/assets/103808587/64779c65-d291-4e4a-b389-63999402fce9)

위 자료에서 볼 수 있듯이, 완전이진트리는 원소들을 배열에 낭비 없이 배치할 수 있습니다. 
배열의 첫번째 인덱스는 계산의 편의를 위해 비워 놓습니다.

완전이진트리를 사용하면 배열에서 다음과 같은 조건을 만족하게 됩니다.

임의의 노드 N이 있을 때,
- `N의 인덱스 * 2 == N의 왼쪽 자식의 인덱스`
- `N의 인덱스 * 2 + 1 == N의 오른쪽 자식의 인덱스`
- `N의 인덱스 / 2 == N의 부모 인덱스`

## 이진 힙의 연산
힙에는 삽입과 삭제, 두 가지 연산이 있습니다. 두 연산 모두 `O(log N)`의 시간 복잡도를 가집니다.

시간 복잡도가 위와 같은 이유는 이진 힙의 연산은 트리의 높이에 의존적이기 때문입니다 
완전 이진 트리의 레벨이 늘어날 때 마다 노드의 수는 2배가 되고, 따라서 트리에 N개의 원소가 있을 때, 트리의 높이는 log N이 되게 됩니다.

편의를 위해 앞으로는 최소 힙을 기준으로 설명하겠습니다.

### 삽입 (Push)
최소 힙의 삽입은 다음과 같은 알고리즘으로 동작합니다.

1. 트리의 가장 끝에 원소를 삽입한다.
2. 그 노드의 부모와 값을 비교해서 만약 부모가 더 작다면 멈춘다.
3. 그렇지 않다면, 부모와 현재 노드를 교환한다. 그리고 2번으로 돌아간다.
 
2번과 3번 단계는 힙 프로퍼티를 만족시키기 위해 트리의 구조를 바꾸는 단계입니다.

![image](https://github.com/subak-ba/cs-study/assets/103808587/8ea9d876-2dec-4a5d-a078-21353437b5a6)

### 삭제 (Pop)
최소 힙의 삭제는 다음과 같은 알고리즘으로 동작합니다.

1. 루트 노드를 제거한다.
2. 트리의 가장 끝에 위치한 노드를 루트 자리에 옮긴다.
3. 만약 옮겨진 노드가 그 자식들보다 작다면 멈춘다.
4. 그렇지 않다면, 더 작은 자식과 노드를 교환한다. 그리고 3번으로 돌아간다.

3번과 4번 단계는 삽입과 마찬가지로 힙 프로퍼티를 만족시키기 위해 트리의 구조를 바꾸는 단계입니다.

![image](https://github.com/subak-ba/cs-study/assets/103808587/33c7fc11-5db0-4552-9414-b03a656d3537)

### 기타
이외에도 힙의 특징을 이용한 몇 가지 연산이 있습니다. 
[이곳](https://en.wikipedia.org/wiki/Binary_heap)에서 더 많은 정보를 찾을 수 있습니다.

## 힙 구현
아래는 golang으로 구현한 힙 자료 구조입니다.

```go
package main

import "fmt"

// 최소 힙. 0번째 인덱스를 비워 놓는다고 가정
type MinHeap struct {
	elems []int // 동적 배열
}

// 삽입
func (h *MinHeap) Push(n int) {
	// 트리의 끝에 노드 삽입
	h.elems = append(h.elems, n)

	// 트리 조정
	h.upheapify(h.length() - 1)
}

// idx에 위치한 노드와 그 부모를 비교하여 swap하는 재귀적인 연산
func (h *MinHeap) upheapify(idx int) {
	// 부모 인덱스
	p := parent(idx)
	if p == 0 {
		// 현재 노드가 루트 노드이므로 끝
		return
	}

	if h.elems[p] <= h.elems[idx] {
		// 부모가 더 작거나 같으므로 끝
		return
	}

	// 교환 후 부모 인덱스에서 다시 검사
	h.swap(idx, p)
	h.upheapify(p)
}

// 삭제
func (h *MinHeap) Pop() int {
	if h.length() == 1 {
		// 엣지케이스
		return -1
	}

	root := h.elems[1]

	// 마지막 노드를 루트 자리에 넣기
	lastNode := h.elems[h.length()-1]
	h.elems[1] = lastNode

	// 배열 길이를 1 줄임
	h.elems = h.elems[:h.length()-1]

	// 트리 조정
	h.downheapify(1)

	return root
}

// idx에 위치한 노드와 그 자식들을 비교하여 swap하는 재귀적인 연산
func (h *MinHeap) downheapify(idx int) {
	// 왼쪽, 오른쪽 자식 인덱스
	lc, rc := leftChild(idx), rightChild(idx)

	if lc < h.length() && h.elems[lc] < h.elems[idx] {
		// 왼쪽 자식이 현재 노드보다 더 작음
		if rc >= h.length() || h.elems[lc] < h.elems[rc] {
			// 왼쪽 자식이 오른쪽 자식보다 더 작기 때문에 스왑 후 이동
			h.swap(idx, lc)
			h.downheapify(lc)
			return
		}
	}
	if rc < h.length() && h.elems[rc] < h.elems[idx] {
		// 오른쪽 자식이 더 작기 때문에 스왑 후 이동
		h.swap(idx, rc)
		h.downheapify(rc)
		return
	}

	// 현재 노드가 자식보다 작음
}

func (h *MinHeap) swap(i, j int) {
	h.elems[i], h.elems[j] = h.elems[j], h.elems[i]
}

func (h *MinHeap) length() int { return len(h.elems) }

func parent(idx int) int     { return idx / 2 }
func leftChild(idx int) int  { return idx * 2 }
func rightChild(idx int) int { return idx*2 + 1 }

func main() {
	heap := MinHeap{[]int{0}} // 0번 인덱스 비우기
	heap.Push(3)
	heap.Push(1)
	heap.Push(2)

	fmt.Println(heap.elems) // [0 1 3 2]

	num := heap.Pop()
	fmt.Println(num) // 1
}
```

## 힙 정렬 (Heap Sort)
힙 정렬은 나중에 정렬 알고리즘 들어갈 때 다시 선택하여 진행하고 링크하겠습니다~~~

### 왜 힙은 이진 힙으로 가장 많이 구현할까?

> because it is simple, cache-friendly, and memory-efficient.

x진 힙의 연산은 x를 상수로 생각했을 때, `O(log N)`의 시간복잡도를 가집니다. 
만약 x를 변수로 생각하면, 시간복잡도는 `O(x * log N / log x)`이 됩니다.

이때, 가장 효율적인 시간복잡도를 찾으려면 `x/log x`을 가장 작게 만드는 x값을 찾으면 됩니다.

![image](https://github.com/subak-ba/cs-study/assets/103808587/06ecfa8e-6389-4167-9bc7-120bb85dc526)

위에서 볼 수 있듯이, 사실은 3진 힙이 가장 효율적인 자료구조라고 합니다.

하지만, 이진 힙을 사용하는 이유는 코드의 단순함과 캐시 지역성, 메모리 효율을 고려했기 때문입니다.

더 자세한 내용은 [스택 오버플로우 원문](https://stackoverflow.com/questions/53970236/why-node-in-heap-data-structure-in-many-types-has-only-two-children)에서 찾아 볼 수 있습니다.

## 레퍼런스
- [나무위키](https://namu.wiki/w/%ED%9E%99%20%ED%8A%B8%EB%A6%AC) (사진)
- [위키피디아: 힙](https://en.wikipedia.org/wiki/Heap_(data_structure))
- [위키피디아: 이진 힙](https://en.wikipedia.org/wiki/Binary_heap)

## 연습문제

우선순위 큐에 대한 문제를 가져왔습니다.

1번 문제: [최소 힙](https://www.acmicpc.net/problem/1927) (실버 2)
개념을 이해했다면 쉽게 풀 수 있는 문제입니다.

2번 문제: [카드 정렬하기](https://www.acmicpc.net/problem/1715) (골드 4)
힙을 이용한 접근을 충분히 고민하면 풀 수 있는 문제입니다.

#17

**Comment by jangwooooo** [comment]
예제 코드가 이해가 잘 되네요. 👍
데이터를 더 넣어서 `downheapify`가 재귀적으로 호출됐으면 더 좋았을 거 같습니다.
연습문제 풀고 추가 코멘트 남기겠습니다 ~
